Add fluid theory

JonathanPham opened the pull request:
## Current situation
Related to issue #33


## Release Notes 
Added theory for fluid portion of svFSIplus


## Documentation
No further documentation needed.


## Testing
No further testing needed.


## Code of Conduct & Contributing Guidelines 
- [x] I agree to follow the [Code of Conduct](https://github.com/SimVascular/.github/blob/main/CODE_OF_CONDUCT.md) and [Contributing Guidelines](https://github.com/SimVascular/.github/blob/main/CONTRIBUTING.md).

JonathanPham committed:
Adding template readme for fluid theory

JonathanPham committed:
Added blank readme for math and gen alpha

JonathanPham committed:
Updated bold symbol

JonathanPham committed:
Updated gen alpha doc

JonathanPham committed:
More gen alpha updates

JonathanPham committed:
More gen alpha updates

JonathanPham committed:
More gen alpha updates

JonathanPham committed:
More gen alpha updates

JonathanPham committed:
More gen alpha updates

JonathanPham committed:
More gen alpha updates

JonathanPham committed:
More gen alpha updates

JonathanPham committed:
More gen alpha updates

JonathanPham committed:
More gen alpha updates

JonathanPham committed:
More gen alpha updates

JonathanPham committed:
More gen alpha updates

JonathanPham committed:
More gen alpha updates

JonathanPham committed:
More gen alpha updates

JonathanPham committed:
More gen alpha updates

JonathanPham committed:
More gen alpha updates

JonathanPham committed:
More gen alpha updates

JonathanPham committed:
More gen alpha updates

JonathanPham committed:
More gen alpha updates

JonathanPham committed:
More gen alpha updates

JonathanPham committed:
More gen alpha updates

JonathanPham committed:
More gen alpha updates

JonathanPham committed:
More gen alpha updates

JonathanPham committed:
More gen alpha updates

JonathanPham committed:
More gen alpha updates

JonathanPham committed:
More gen alpha updates

JonathanPham committed:
More gen alpha updates

JonathanPham committed:
More gen alpha updates

JonathanPham committed:
More gen alpha updates

JonathanPham committed:
More gen alpha updates

JonathanPham committed:
More gen alpha updates

JonathanPham committed:
Updated fluid theory

JonathanPham committed:
Updated fluid theory

JonathanPham committed:
Updated fluid theory

JonathanPham committed:
Updated fluid theory

JonathanPham committed:
Updated fluid theory

JonathanPham committed:
Updated fluid theory

JonathanPham committed:
Updated fluid theory

JonathanPham committed:
Updated fluid theory

JonathanPham committed:
Updated fluid theory

JonathanPham committed:
Updated fluid theory

JonathanPham committed:
Updated fluid theory

JonathanPham committed:
Updated fluid theory

JonathanPham committed:
Updated fluid theory

JonathanPham committed:
Updated fluid theory

JonathanPham committed:
Updated fluid theory

JonathanPham committed:
Updated fluid theory

JonathanPham committed:
Updated fluid theory

JonathanPham committed:
Updated fluid theory

JonathanPham committed:
Updated fluid theory

JonathanPham committed:
Updated fluid theory

JonathanPham committed:
Updated fluid theory

JonathanPham committed:
Updated fluid theory

JonathanPham committed:
Updated fluid theory

JonathanPham committed:
Updated fluid theory

JonathanPham committed:
Updated fluid theory

JonathanPham committed:
Updated fluid theory

JonathanPham committed:
Updated fluid theory

JonathanPham committed:
Updated fluid theory

JonathanPham committed:
Updated fluid theory

JonathanPham committed:
Updated fluid theory

JonathanPham committed:
Updated fluid theory

JonathanPham committed:
Updated fluid theory

JonathanPham committed:
Updated fluid theory

JonathanPham committed:
Updated fluid theory

JonathanPham committed:
Updated fluid theory

JonathanPham committed:
Updated fluid theory

JonathanPham committed:
Updated fluid theory

JonathanPham committed:
Updated fluid theory

JonathanPham committed:
Updated fluid theoryg

JonathanPham committed:
Updated fluid theory

JonathanPham committed:
Updated fluid theory

JonathanPham committed:
Updated fluid theory

JonathanPham committed:
Updated fluid theory

JonathanPham committed:
Updated fluid theory

JonathanPham committed:
Updated fluid theory

JonathanPham committed:
Updated fluid theory

JonathanPham committed:
Updated fluid theory

JonathanPham committed:
Updated fluid theory

JonathanPham committed:
Updated fluid theory

JonathanPham committed:
Updated fluid theory

JonathanPham committed:
Updated fluid theory

JonathanPham committed:
Updated fluid theory

JonathanPham committed:
Updated fluid theory

JonathanPham committed:
Updated fluid theory

JonathanPham committed:
Updated fluid theory

JonathanPham committed:
Updated fluid theory

JonathanPham committed:
Updated fluid theory

JonathanPham committed:
Updated fluid theory

JonathanPham committed:
Updated fluid theory

JonathanPham committed:
Updated fluid theory

JonathanPham committed:
Updated fluid theory

JonathanPham committed:
Updated fluid theory

JonathanPham committed:
Updated fluid theory

JonathanPham committed:
Updated fluid theory

JonathanPham committed:
Updated fluid theory

JonathanPham committed:
Updated fluid theory

JonathanPham committed:
Updated fluid theory

JonathanPham committed:
Updated fluid theory

JonathanPham committed:
Updated fluid theory

JonathanPham committed:
Updated fluid theory

JonathanPham committed:
Updated fluid theory

JonathanPham committed:
Updated fluid theory

JonathanPham committed:
Updated fluid theory

JonathanPham committed:
Updated fluid theory

JonathanPham committed:
Updated fluid theory

JonathanPham committed:
Updated fluid theory

JonathanPham committed:
Updated fluid theory

JonathanPham committed:
Updated fluid theory

JonathanPham committed:
Updated fluid theory

JonathanPham committed:
Updated fluid theory

JonathanPham committed:
Updated fluid theory

JonathanPham committed:
Updated fluid theory

JonathanPham committed:
Updated fluid theory

JonathanPham committed:
Updated fluid theory

JonathanPham committed:
Updated fluid theory

JonathanPham committed:
Updated fluid theory

JonathanPham committed:
Updated fluid theory

JonathanPham committed:
Updated fluid theory

JonathanPham committed:
Updated fluid theory

JonathanPham committed:
Updated fluid theory

JonathanPham committed:
Updated fluid theory

JonathanPham committed:
Updated fluid theory

JonathanPham committed:
Updated fluid theory

JonathanPham committed:
Updated fluid theory

JonathanPham committed:
Updated fluid theory

JonathanPham committed:
Updated fluid theory

JonathanPham committed:
Updated fluid theory

JonathanPham committed:
Updated fluid theory

JonathanPham committed:
Updated fluid theory

JonathanPham committed:
Updated fluid theory

JonathanPham committed:
Updated fluid theory

JonathanPham committed:
Updated fluid theory

JonathanPham committed:
Updated fluid theory

JonathanPham committed:
Updated fluid theory

JonathanPham committed:
Updated fluid theory

JonathanPham committed:
Updated fluid theory

JonathanPham committed:
Updated fluid theory

JonathanPham committed:
Updated fluid theory

JonathanPham committed:
Updated fluid theory

JonathanPham committed:
Updated fluid theory

JonathanPham committed:
Updated fluid theory

JonathanPham committed:
Updated fluid theory

JonathanPham committed:
Updated fluid theory

JonathanPham committed:
Updated fluid theory

JonathanPham committed:
Updated fluid theory

JonathanPham committed:
Updated fluid theory

JonathanPham committed:
Updated fluid theory

JonathanPham committed:
Updated fluid theory

JonathanPham committed:
Updated fluid theory

JonathanPham committed:
Updated fluid theory

JonathanPham committed:
Updated fluid theory

JonathanPham committed:
Updated fluid theory

JonathanPham committed:
Updated fluid theory

JonathanPham committed:
Updated fluid theory

JonathanPham committed:
Updated fluid theory

JonathanPham committed:
Updated fluid theory

JonathanPham committed:
Updated fluid theory

JonathanPham committed:
Updated fluid theory

JonathanPham committed:
Updated fluid theory

JonathanPham committed:
Updated fluid theory

JonathanPham committed:
Updated fluid theory

JonathanPham committed:
Updated fluid theory

JonathanPham committed:
Updated fluid theory

JonathanPham committed:
Updated fluid theory

JonathanPham committed:
Updated fluid theory

JonathanPham committed:
Updated fluid theory

JonathanPham committed:
Updated fluid theory

JonathanPham committed:
Updated fluid theory

JonathanPham committed:
Updated fluid theory

JonathanPham committed:
Updated fluid theory

JonathanPham committed:
Updated fluid theory

JonathanPham committed:
Updated fluid theory

JonathanPham committed:
Updated fluid theory

JonathanPham committed:
Updated fluid theory

JonathanPham committed:
Updated fluid theory

JonathanPham committed:
Updated fluid theory

ktbolt commented:
@JonathanPham A good start but I am seeing some problems with 

`1) pacing and trailing commas`
<img width="600" alt="trailing commas and spacing" src="https://github.com/user-attachments/assets/60c10b45-6f6d-4fc9-94e7-c7e97a83987b">


`2) equation is not being rendered correctly`
<img width="630" alt="eq-error" src="https://github.com/user-attachments/assets/469c40a6-5928-4330-86e5-fa714e50c0f6">

I don't think you need to use `\begin{aligned}`


The spacing problem is I think caused by ` &= 0`

JonathanPham committed:
Updated fluid theory

JonathanPham committed:
Updated fluid theory

JonathanPham committed:
Updated fluid theory

JonathanPham committed:
Updated fluid theory

JonathanPham committed:
Updated fluid theory

JonathanPham committed:
Updated fluid theory

JonathanPham commented:
I've removed `\begin{aligned}` and ` &= 0`. I think the math renders correctly now:

- <img width="244" alt="Screenshot 2024-10-30 at 1 43 06 PM" src="https://github.com/user-attachments/assets/9a9b3b33-bbae-4452-a3f4-59652800388c">
- <img width="279" alt="Screenshot 2024-10-30 at 1 43 11 PM" src="https://github.com/user-attachments/assets/70fccfdb-203c-47c1-979d-18da92c784b4">

ktbolt reviewed (commented):
I would start off with describing the Navier-Stokes equation in terms of just velocity and pressure, not the Navier-Stokes-Brinkman equation.

ktbolt commented:
@dcodoni @aabrown100-git @msbazzi Please have a look at the documentation.

ktbolt commented:
@JonathanPham I think you should add a bit more background information to each section. The users reading this documentation may have little or no finite element background.

It will be helpful to described the theory and meaning of things users will encounter when setting up and running simulations. The **Residuals** section is good because this is something users will see in the svFSIplus log file.

ktbolt commented:
@JonathanPham References ?

JonathanPham committed:
Navier-Stokes first before Brinkman

JonathanPham committed:
Adding reference template, math preliminaries, and gen alpha

JonathanPham committed:
Added fluid references

JonathanPham committed:
Update gen alpha

JonathanPham committed:
Removed gen alpha

JonathanPham committed:
Updated doc

JonathanPham commented:
I've added references and a bit more background. Also, updated doc to discuss Navier-Stokes first, before introducing the Brinkman. 

However, I think an overview of the finite element method and time integration should go in its own section, since it is pertinent to the other physics modules in svFSIplus, not only the fluids module.

ktbolt commented:
@JonathanPham Looks good!

There is an brief overview of the finite element method in the `Overview` section but that needs more detail, maybe there should be a separate section for that, other Lab member should contribute.

dcodoni commented on documentation/svfsiplus/theory/fluid/readme.md:
```diff
@@ -0,0 +1,258 @@
+$$
+
+$$
+\sigma_{ij,j} = -p_{,i} + 2\epsilon_{ij}\frac{\partial \mu}{\partial x_{j}} + \mu\nabla^{2}u_{i},
```
@JonathanPham  Here you are writing the divergence of sigma in index notation. For consistency shouldn't the laplacian operator be rewritten in index notation too?

JonathanPham replied:
Good idea, done!

dcodoni commented on documentation/svfsiplus/theory/fluid/readme.md:
```diff
@@ -0,0 +1,258 @@
+For the finite element method, we will first derive the Galerkin weak form for the Navier-Stokes-Brinkman equations. We define our trial and weighting function spaces,
+
+$$
+u_{i} \in \tau_{i} = \{u_{i} \in H^{1}\left(\Omega\right) \mid u_{i} = g_{i} \text{ on } \Gamma_{g_{i}}\},
```
@JonathanPham I am used to define the function spaces in {} and not using =, I find it more clear. For example: u \in V:{u\in H^1| u=g}. This is my personal preference though.

JonathanPham replied:
Makes sense. I originally had the brackets, but they didn't render correctly. Fixed the issue now.

dcodoni commented on documentation/svfsiplus/theory/fluid/readme.md:
```diff
@@ -0,0 +1,258 @@
+We then multiply the Navier-Stokes-Brinkman equations by $\boldsymbol{w}$ and $q$, respectively, and integrate by parts to obtain the standard Galerkin momentum and continuity weak forms <a href="#ref-7">[7]</a> <a href="#ref-5">[5]</a>,
+
+$$
+\int \rho w_{i}\frac{du_{i}}{dt} \,d\Omega + \int \rho w_{i}u_{k}u_{i, k} \,d\Omega + \int w_{i, j}\sigma_{ij} \,d\Omega + \int \frac{\mu}{K}w_{i}u_{i} \,d\Omega - \int w_{i}\rho b_{i} \,d\Omega - \int_{\Gamma_{h}} w_{i}h_{i} \,d\Gamma = 0,
```
@JonathanPham I would add _\Omega for each volumetric integral for clarity.

JonathanPham replied:
Good idea.

dcodoni commented on documentation/svfsiplus/theory/fluid/readme.md:
```diff
@@ -0,0 +1,258 @@
+The stabilization parameters are defined as
+
+$$
+\tau_{SUPS} = \tau_{M} = \left(\frac{4}{\Delta t^{2}} + \boldsymbol{u}^{h} \cdot \boldsymbol{G}\boldsymbol{u}^{h} + C_{1}\nu^{2}\boldsymbol{G}:\boldsymbol{G} + \left(\frac{\nu}{K}\right)^{2}\right)^{-1/2},
```
@JonathanPham do you think we should define G as the metric tensor or is it too much detail?

JonathanPham replied:
Just added that and a reference too.

dcodoni commented on documentation/svfsiplus/theory/fluid/readme.md:
```diff
@@ -0,0 +1,258 @@
+\int qu_{i,i} \,d\Omega + \int \tau_{SUPS}\frac{q_{,i}}{\rho}r_{Mi} \,d\Omega = 0.
+$$
+
+Then, by plugging weighting functions into these equations and holding the results true for any arbitrary $w_{ai}$ and $q_{a}$, we obtain the per-element momentum and continuity residuals,
```
@JonathanPham I think these residuals are per-node and not per-element. It may be confusing saying per-element, since we solve per node.

dcodoni reviewed (commented):
@JonathanPham  well done!
I left few comments, very minor modifications in case you think they are useful.

JonathanPham committed:
Laplacian in index notation

JonathanPham committed:
Laplacian in index notation

JonathanPham committed:
Brackets for function spaces

JonathanPham committed:
Brackets for function spaces

JonathanPham committed:
Brackets for function spaces

JonathanPham committed:
Brackets for function spaces

JonathanPham committed:
Brackets for function spaces

JonathanPham committed:
Brackets for function spaces

JonathanPham committed:
Brackets for function spaces

JonathanPham committed:
Brackets for function spaces

JonathanPham committed:
Add Omega for volume integrals

JonathanPham committed:
Add Omega for volume integrals

JonathanPham committed:
Add Omega for volume integrals

JonathanPham committed:
Add Omega for volume integrals

JonathanPham committed:
Add Omega for volume integrals

JonathanPham committed:
Add Omega for volume integrals

JonathanPham committed:
Add Omega for volume integrals

JonathanPham committed:
Add Omega for volume integrals

JonathanPham committed:
Add Omega for volume integrals

JonathanPham committed:
Add Omega for volume integrals

JonathanPham committed:
Add Omega for volume integrals

JonathanPham committed:
Add Omega for volume integrals

JonathanPham committed:
Add Omega for volume integrals

JonathanPham committed:
Add Omega for volume integrals

JonathanPham committed:
Add Omega for volume integrals

JonathanPham committed:
Add Omega for volume integrals

JonathanPham committed:
Add Omega for volume integrals

JonathanPham committed:
Add Omega for volume integrals

JonathanPham committed:
Add Omega for volume integrals

JonathanPham committed:
Add Omega for volume integrals

JonathanPham committed:
Add Omega for volume integrals

JonathanPham committed:
Add Omega for volume integrals

JonathanPham committed:
residuals not at elements

JonathanPham committed:
residuals not at elements

JonathanPham commented:
Thanks @ktbolt and @dcodoni for the suggestions. Update the doc to reflect your feedback. Let me know if we are good to merge this pull request now.

dcodoni commented:
Thanks @JonathanPham !
